### PR TITLE
Implement checkstyle rules for App Inventor

### PR DESCRIPTION
**Pull request by ewpatton**
This commit implements checkstyle for App Inventor's Java code. The
App Inventor style rules are based on the Google style rules, with an
exception added for methods in the components/runtime package that are
annotated with `@Simple*` or `@Design*` annotations. The App Inventor
specific rules are placed at the end of the appinventor_checks.xml
file.

In addition, there is a script called checkstyle_git.py that is
written in Python 3 and will check the git diff against the rules by
running checkstyle on the changed files and crossreferencing the line
numbers with the checkstyle output. This can be used as a git
pre-commit hook to check that changes conform to the style guide. The
script returns -1 on error, which will abort the commit until the
changes confirm. This behavior can be bypassed by exporting the
`BYPASS_CHECKSTYLE` environment variable.

These two changes are made available as ant targets:

- `ant checkstyle`: Run the checkstyle_git.py script to check the
  git diff for problems.
- `ant checkstyle-all`: Run checkstyle on the whole repo.

Change-Id: I26c6049bd49d1ce3f9b8dbd48f8f115a89fb5b52

### CODE REVIEW

**Comment by SusanRatiLane** [comment]
My first thought is that checkstyle should be the one that checks git diff, and the target that checks the entire project should be something like checkstyle-all. We're not going to use the check against the entire project very often.

It would also be nice to have a target that checks staged-but-not committed files, if that is possible. It would allow us to do a spot-check before making a commit.

**Comment by ewpatton** [comment]
@SusanRatiLane The `checkstyle-git` target will handle staged but not committed files. I have a Java file I used to test the rules called CheckstyleTest.java. If I stage it and then run `ant checkstyle-git`, it will fail the build.

components/src/com/google/appinventor/components/runtime/CheckstyleTest.java:

```java
package com.google.appinventor.components.runtime;

import com.google.appinventor.components.annotations.DesignerProperty;
import com.google.appinventor.components.annotations.SimpleEvent;
import com.google.appinventor.components.annotations.SimpleFunction;
import com.google.appinventor.components.annotations.SimpleProperty;

public class CheckstyleTest extends AndroidNonvisibleComponent {
  /**
   * Creates a new AndroidNonvisibleComponent.
   *
   * @param form the container that this component will be placed in
   */
  public CheckstyleTest(Form form) {
    super(form);
  }

  public void Initialize() {}

  @Deprecated
  public void BadMethod() {}

  @SimpleProperty
  public void AppInventorSimpleProperty(String foo) {}

  @DesignerProperty
  public void AppInventorDesignerProperty(String foo) {}

  @SimpleFunction
  public void AppInventorFunction() {}

  @SimpleEvent
  public void AppInventorEvent() {}

  public void regularMethod() {}

  @Deprecated
  @SimpleEvent
  public void badRegularMethod() {}

  @DesignerProperty
  public void badRegularFunction() {}
}
```

**Comment by ewpatton** [comment]
@SusanRatiLane changed the target names per your comment.

**Comment by SusanRatiLane** [comment]
> @SusanRatiLane The `checkstyle-git` target will handle staged but not committed files. I have a Java file I used to test the rules called CheckstyleTest.java. If I stage it and then run `ant checkstyle-git`, it will fail the build.

Hmm. Perhaps I don't have a clear enough idea of what the style checks are looking for. I made inappropriate whitespace changes to a file, staged it, and ran checkstyle-git. It reported success.

**Comment by ewpatton** [comment]
@SusanRatiLane Hmm... Looks like a potential bug in my script. I followed your procedure and it does indeed pass, but if I run checkstyle-all I can see the erroneous lines in the output. The style rules should be correct though.

**Comment by ewpatton** [comment]
@SusanRatiLane I've pushed a new version of the checkstyle-git script. It should address the problems you encountered.

**Comment by SusanRatiLane** [comment]
@ewpatton All checks out now. Great.

I don't know if this is a good place to debate style rules, but I have to say that I think the 100 character length limit is incredibly short for modern screen resolution. With fairly long java method and class names, I think the line limit makes things harder to read rather than easier. Is this something we could adjust? Can I convince anyone else to adjust it? :)

**Comment by pavi2410** [comment]
Can CheckStyle also catch errors when defining a `@SimpleEvent` method with wrong method name in the `EventDispatcher.dispatchEvent()`?

File: appinventor/misc/checkstyle/checkstyle-git.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    :param checkstyle_output:
+    :param chunks:
+    :type chunks: list[(int, int)]
+    :return:
+    """
+    success = True
+    i = 0
+    for line in checkstyle_output:
+        match = checkstyle_pattern.match(line.strip())
+        if match:
```
Comment (pavi2410) [inline appinventor/misc/checkstyle/checkstyle-git.py:34]: This line can use the new Walrus operator `:=` introduced with Python 3.8

https://www.python.org/dev/peps/pep-0572/

---
You can choose to ignore this one but I couldn't resist 😄

**Comment by ewpatton** [comment]
@pavi2410 The commit I just pushed adds some additional checks (implemented in Java) that will check the following three App Inventor conditions:

1. `@SimpleEvent` annotated methods call EventDispatcher.dispatchEvent (EventDispatchMissing)
2. `@SimpleEvent` annotated methods use the correct name in the second parameter (EventDispatchWrongName)
3. `@SimpleEvent` annotated methods use the right number of arguments in the call (EventDispatchArguments)

**Comment by ewpatton** [comment]
> I don't know if this is a good place to debate style rules, but I have to say that I think the 100 character length limit is incredibly short for modern screen resolution. With fairly long java method and class names, I think the line limit makes things harder to read rather than easier. Is this something we could adjust? Can I convince anyone else to adjust it? :)

@SusanRatiLane I'd rather not make any deviations from the existing rules right now. As a counterpoint, I often split my windows half and half on the screen so the 100 char limit width-wise is actually useful there. It's also going to be a function of a number of other factors, including font size and overall display size. I also sometimes view the patches on my phone (because I'm weird like that).

**Comment by kkashi01** [comment]
@ewpatton this is a great check and will reduce errors

**Pull request merged**